### ISSUE TITLE
Add a way to handle MultiTenantAuthenticationService.ChallengeAsync error

### ISSUE THREAD

**Issue by natelaff** [Finbuckle/Finbuckle.MultiTenant#349]
I've been dealing with this for a little while and finally went to fix it, but upon research I don't think there is a good way to handle this scenario without a bit of a change to this package.

I use OpenIdConnect and point to a tenant aware IdentityServer instance which uses tenant name in the path.

I have my HomeController with an AuthorizeAttribute on it. So what ends up happening here is if I put in an invalid tenant name, it immediately tries to challenge OpenIdConnect and fails with the below exception because ChallengeAsync is happening before the tenant is actually resolved on the client. So I'm kind of at a loss for how to gracefully handle this situation.

**InvalidOperationException: Cannot redirect to the authorization endpoint, the configuration may be missing or invalid.**

Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectHandler.HandleChallengeAsyncInternal(AuthenticationProperties properties)
Microsoft.AspNetCore.Authentication.OpenIdConnect.OpenIdConnectHandler.HandleChallengeAsync(AuthenticationProperties properties)
Microsoft.AspNetCore.Authentication.AuthenticationHandler<TOptions>.ChallengeAsync(AuthenticationProperties properties)
Microsoft.AspNetCore.Authentication.AuthenticationService.ChallengeAsync(HttpContext context, string scheme, AuthenticationProperties properties)
Finbuckle.MultiTenant.AspNetCore.MultiTenantAuthenticationService<TTenantInfo>.ChallengeAsync(HttpContext context, string scheme, AuthenticationProperties properties)
Microsoft.AspNetCore.Authorization.AuthorizationMiddleware.Invoke(HttpContext context)
Microsoft.AspNetCore.Authentication.AuthenticationMiddleware.Invoke(HttpContext context)
Finbuckle.MultiTenant.AspNetCore.MultiTenantMiddleware.Invoke(HttpContext context)
Microsoft.AspNetCore.Diagnostics.DeveloperExceptionPageMiddleware.Invoke(HttpContext context)

**Comment by AndrewTriesToCode** [comment]
@natelaff 
I've run into this myself. After examining the [OpenIdConnectHandler](https://github.com/dotnet/aspnetcore/blob/master/src/Security/Authentication/OpenIdConnect/src/OpenIdConnectHandler.cs) I found that this is happening because it's unable to get the OpenID Connect configuration from the service discovery endpoint.

In my case it seems that IdentityServer didn't want to play nice with custom routes, especially for the service discovery endpoint. I.e. `https://server.com/.well-known/openid-configuration` was the only reliable method.

For the authorization request they actually [support a tenant parameter](https://identityserver4.readthedocs.io/en/latest/reference/interactionservice.html#authorizationrequest) which can be used in a custom tenant strategy to get the tenant for that request. I use that to get the user logged in, and from then on use the ClaimStrategy (ensuring the generated user had a tenant claim) for subsequent requests on the IdentityServer host and things works out pretty well.

This is getting out of basic sample project territory, but I do plan a few blog posts and a sample app on this stuff in the near future as premium content for github sponsors (coming soon).

**Comment by natelaff** [comment]
Yes, when I looked I found it silly that it just throws from AspNetCore there instead of raises one of the OpenIdConnect failure events. That's where I figured trapping it within Finbuckle could be a good solution, albeit perhaps out of scope, but yet, in a way, also in scope since you're allowing dynamic endpoints through.

Your method of tossing it in the claim after login then checking from that point on is good, yet you lose a pretty decent extra security check by doing so which is that the authority matches the tenant name. How important this _really_ is, hard to say, but it certainly doesn't hurt anything and I've found it to be a nice addition.
